app/models/container_label_tag_mapping.rb

```diff
@@ -9,91 +9,82 @@ class ContainerLabelTagMapping < ApplicationRecord
+     map_name_type_value(cache, label[:name], nil,  label[:value]))
   end
 
-  def self.tags_for_name_type_value(name, type, value)
-    specific_value = hash_all_by_name_type_value[[name, type, value]] || []
-    any_value      = hash_all_by_name_type_value[[name, type, nil]]   || []
+  def self.map_name_type_value(cache, name, type, value)
+    byebug if cache.nil?
+    specific_value = cache[[name, type, value]] || []
+    any_value      = cache[[name, type, nil]]   || []
     if !specific_value.empty?
-      specific_value
+      specific_value.map { |tag_id| {:tag_id => tag_id} }
     else
-      any_value.map do |category_tag|
-        create_specific_value_mapping(name, type, value, category_tag).tag
+      if value.empty?
+        [] # Don't map empty value to any tag.
+      else
+        # Note: if the way we compute `entry_name` changes,
+        # consider what will happen to previously created tags.
+        any_value.map do |tag_id|
+          {:category_tag_id   => tag_id,
+           :entry_name        => Classification.sanitize_name(value),
+           :entry_description => value}
+        end
       end
     end
   end
-  private_class_method :tags_for_name_type_value
+  private_class_method :map_name_type_value
 
-  # If this is an open ended any-value mapping, finds or creates a
-  # specific-value mapping to a specific tag.
-  def self.create_specific_value_mapping(name, type, value, category_tag)
-    new_tag = create_tag(name, value, category_tag)
-    new_mapping = create!(:labeled_resource_type => type, :label_name => name, :label_value => value,
-                          :tag => new_tag)
-    load_mapping_into_hash(new_mapping)
-    new_mapping
-  end
-  private_class_method :create_specific_value_mapping
-
-  def self.create_tag(name, value, category_tag)
-    category = category_tag.classification
-    unless category
-      category = Classification.create_category!(:description => "Kubernetes label '#{name}'",
-                                                 :read_only   => true,
-                                                 :tag         => category_tag)
-    end
-
-    if value.empty?
-      entry_name = ':empty:' # ':' character won't occur in kubernetes values.
-      description = '<empty value>'
+  # Given a hash built by `map_*` methods, returns a Tag (creating if needed).
+  def self.find_or_create_tag(tag_hash)
+    if tag_hash[:tag_id]
+      Tag.find(tag_hash[:tag_id])
     else
-      entry_name = Classification.sanitize_name(value)
-      description = value
+      category = Tag.find(tag_hash[:category_tag_id]).classification
+      entry = category.find_entry_by_name(tag_hash[:entry_name])
+      unless entry
+        category.lock :exclusive do
+          begin
+            entry = category.add_entry(:name => tag_hash[:entry_name],
+                                       :description => tag_hash[:entry_description])
+            entry.save!
+          rescue ActiveRecord::RecordInvalid
```

Comment:
This is not DB error (that would be `ActiveRecord::RecordNotUnique`).
This is save! failing Rails' `validates :uniqueness`.  I believe it can happen if other process created it between first `find_entry_by_name` and taking the lock.
(I want to keep first lookup outside of the lock section so I can avoid locking entirely in the happy path.)

[Everything I say about locking and races should be double-checked as I'm new to this]